### ISSUE TITLE
std in MSGD-Fisher

### ISSUE THREAD

**Issue by XuanyuWu123** [uuujf/MultiNoise#1]
Hi, I read the code and notice that the default value for std in MSGD-Fisher is 0 and there is no calculation in the code to replace it. I assume std of 0 is not what the paper proposes because that would make the mask all ones. How do you calculate std in your experiments? Can you upload code for that part or the idea of how to implement it? Thanks.

**Comment by uuujf** [comment]
> Hi, I read the code and notice that the default value for std in MSGD-Fisher is 0 and there is no calculation in the code to replace it. I assume std of 0 is not what the paper proposes because that would make the mask all ones. How do you calculate std in your experiments? Can you upload code for that part or the idea of how to implement it? Thanks.

Hi, sorry for the ambiguous description in the code. The key vagueness comes from the tricky (but standard) implementation of function `CEwithMask`  within `utils.py`. 

Here is some idea how to compute sigma:

We always set mean to be 1. Suppose the batch size is n (full batch), and noise is eps, then the data weight, i.e., mask, is (1+eps); but when applied in `CEwithMask`, we call function `mean`, thus the true weight is (1+\eps)/n, then we compute the variance as 

variance = E [ 1/n * n * ( (1+eps)/n - 1/n  )^2 ] = E [ eps^2 ] / n = std^2 / n

Then we obtain eps according to the variance used in the paper (supplementary materials).

Note:
You may also use 1/(n-1) instead of 1/n to estimate the variance, but you would not lose much use the above formula.

**Comment by XuanyuWu123** [comment]
Hi, thanks for your response. But I still have some confusions. 

I feel like you start with eps to get variance, but in the end you mention to obtain eps according to variance, which sounds like a cycle...

From my understanding, in your implementation, you generate eps and mask by inputting args.std, whose default value is 0. Does that mean we need to precompute args.std and input it as an argument? Can we calculate it based on dataset/model?

> variance = E [ 1/n * n * ( (1+eps)/n - 1/n )^2 ] = E [ eps^2 ] / n = std^2 / n

Which variance are you referring to? Is it the variance of eps?

> Then we obtain eps according to the variance used in the paper (supplementary materials).

Can you specify where in the supplementary materials can I find how to obtain eps? Because I can only see proofs and general experimental settings.

Some clarifications would really help. Thanks!

**Comment by uuujf** [comment]
Hi,

1. Our work shows that, if one already knows a small batch SGD works, one can calculate a std so that large batch SGD with multiplicative noise also works under the same dataset/model/hyperparameters. However, we are unable to pre-determine the best std for MSGD based on dataset/model.

2. I would interpret std as a hyperparameter for MSGD. Typically it requires tuning; but if you already know some SGD configuration works (e.g., batch size + learning rate), you can obtain a std for MSGD on the fly based on our paper.

3. The variance I mentioned means the variance for the multiplicative weights in small batch SGD, which can be obtained by Proposition 1 in our paper (of course, you need batch size b and number of data n). Then we calculate a std so that the variance of weights in MSGD approximately matches that of small batch SGD.

Please feel free to discuss more if I have not made things clear :)
